stages/org.osbuild.coreos.platform

```diff
@@ -0,0 +1,115 @@
+#!/usr/bin/python3
+"""
+In CoreOS we have the concept of a platform (i.e. AWS, GCP, Metal, QEMU)
+where each platform has its own provided disk image with slightly
+differing settings/behavior. This stage will perform the necessary
+configuration for the given platform. This configuration boils down to
+a few steps:
+1. Locate the source of platform specific information that is provided
+   in the CoreOS filesystem tree already (the platforms.json).
+2. Copy the platforms.json file into the /boot/ partition, which is
+   sometimes used by coreos-installer.
+3. Read the platforms.json to fetch and platform specific kernel
+   arguments or grub configuration to set. These arguments/config
+   are primarily console settings.
+4. Apply any platform specific kernel arguments along with the
+   `ignition.platform.id={platform-name}` kernel argument.
+5. Create the grub console.cfg file and apply any platform
+   specific grub console configuration.
+This stage is highly CoreOS specific and subject to change in the
+future if/when we change the way platform specific information is
+defined in our broader efforts to share more defaults with OSBuild.
+"""
+SCHEMA_2 = r"""
+"options": {
+  "additionalProperties": false,
+  "properties": {
+    "platform": {
+      "description": "The target platform name/ID",
+      "type": "string"
+    }
+  }
+},
+"devices": {
+  "type": "object",
+  "additionalProperties": true
+},
+"mounts": {
+  "type": "array"
+}
+"""
+import json
+import os
+import osbuild.api
+import shutil
+import sys
+
+from osbuild.util import bls
+
+# Constants
+#For console.cfg see https://github.com/coreos/bootupd/pull/620
+#For platforms.json see https://github.com/coreos/coreos-assembler/pull/3709
+GRUB_CONSOLE_CFG = "boot/grub2/console.cfg"
+PLATFORMS_JSON_DEST= "boot/coreos/platforms.json"
+PLATFORMS_JSON_SOURCE = "usr/share/coreos-assembler/platforms.json"
+
+def generate_console_settings_file(console_settings, file_path):
+    settings_content = ""
+    if console_settings is not None:
+        settings_content = "\n".join(console_settings)
+    file_content = f"""
+# Any non-default console settings will be inserted here.
+# CONSOLE-SETTINGS-START
+{settings_content}
+# CONSOLE-SETTINGS-END
+"""
+    with open(file_path, 'w') as file:
+        file.write(file_content)
+
+
+def process_platforms_json(json_file_path, platform):
+    keys = ["grub_commands", "kernel_arguments"]
+    result = {}
+    with open(json_file_path, 'r') as file:
```

Comment:
Is there a link to an example json input? I would love to have a test but my gut feeling is also that this could be slightly simplified and the json input as a string would really help me :)

Reply:
here's an example:

```
[dustymabe@media ~]$ podman run --rm quay.io/fedora/fedora-coreos:stable cat /usr/share/coreos-assembler/platforms.json 
{
  "aws": {
    "grub_commands": [
      "serial --speed=115200",
      "terminal_input serial console",
      "terminal_output serial console"
    ],
    "kernel_arguments": [
      "console=tty0",
      "console=ttyS0,115200n8"
    ]
  },
  "azure": {
    "kernel_arguments": [
      "console=tty0",
      "console=ttyS0,115200n8"
    ]
  },
  "gcp": {
    "grub_commands": [
      "serial --speed=115200",
      "terminal_input serial console",
      "terminal_output serial console"
    ],
    "kernel_arguments": [
      "console=tty0",
      "console=ttyS0,115200n8"
    ]
  },
  "ibmcloud": {
    "grub_commands": [
      "serial --speed=115200",
      "terminal_input serial console",
      "terminal_output serial console"
    ],
    "kernel_arguments": [
      "console=tty0",
      "console=ttyS0,115200n8"
    ]
  },
  "kubevirt": {
    "grub_commands": [
      "serial --speed=115200",
      "terminal_input serial console",
      "terminal_output serial console"
    ],
    "kernel_arguments": [
      "console=ttyS0,115200n8",
      "console=tty0"
    ]
  },
  "openstack": {
    "grub_commands": [
      "serial --speed=115200",
      "terminal_input serial console",
      "terminal_output serial console"
    ],
    "kernel_arguments": [
      "console=ttyS0,115200n8",
      "console=tty0"
    ]
  },
  "packet": {
    "grub_commands": [
      "serial --unit=1 --speed=115200",
      "terminal_input serial",
      "terminal_output serial"
    ],
    "kernel_arguments": [
      "console=ttyS1,115200n8"
    ]
  },
  "qemu": {
    "grub_commands": [
      "serial --speed=115200",
      "terminal_input serial console",
      "terminal_output serial console"
    ],
    "kernel_arguments": [
      "console=tty0",
      "console=ttyS0,115200n8"
    ]
  },
  "virtualbox": {
    "grub_commands": [
      "serial --speed=115200",
      "terminal_input serial console",
      "terminal_output serial console"
    ],
    "kernel_arguments": [
      "console=ttyS0,115200n8",
      "console=tty0"
    ]
  },
  "vmware": {
    "grub_commands": [
      "serial --speed=115200",
      "terminal_input serial console",
      "terminal_output serial console"
    ],
    "kernel_arguments": [
      "console=ttyS0,115200n8",
      "console=tty0"
    ]
  }
}

```

Reply:
the source for this ultimately comes from https://github.com/coreos/fedora-coreos-config/blob/testing-devel/platforms.yaml which then gets converted to json and then stripped down to the architecture of the artifact we are building

Reply:
Thanks, this is super helpful. I wrote a test and also tweaked the helper a tiny bit (removing essentially the intermediate result struct that is nice for more inputs but given that we just have two keys seems to be not quite needed (of course fine if you folks disagree and we can revert/remove the tweak, the tests is what is important to me :)

Reply:
will review when we open the followup PR